### PR TITLE
Prevent FlashInfer sampling from running with CUDA graphs

### PR DESCRIPTION
**Pull request by santhnm2**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Prevents FlashInfer sampling from running with CUDA graphs enabled due to a known issue regarding the RNG state not being CUDA-graph safe.
See here for more details: https://www.linkedin.com/pulse/pinned-rng-drifting-crash-from-cuda-graph-chenyang-zhao-csuac/

Also enables more fine-grained dispatching across sampling kernels depending on the batch composition (i.e., different kernels for only top-k, only top-p, and mixed batches).

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by kvareddy** [comment]
@tdene can you please take a look at this MR?

File: megatron/core/inference/sampling/flashinfer_sampling.py [hunk 1]
```diff
@@ -79,25 +84,56 @@ def sample_kernel(
...
+        # philox offset advances per launch.
         output = torch.empty(n, device=logits.device, dtype=torch.int64)
-        output.copy_(
-            flashinfer.sampling.top_k_top_p_sampling_from_probs(
-                probs, top_k_safe, top_p_safe, generator=self._rng
+        if no_top_k and no_top_p:
+            # No nucleus / top-k filtering: sample the full temperature-scaled
+            # distribution. (torch.multinomial keeps this rare branch dependency-free.)
+            probs = torch.softmax(scaled, dim=-1)
+            output.copy_(torch.multinomial(probs, num_samples=1, generator=self._rng).view(-1))
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/sampling/flashinfer_sampling.py]: torch.multinomial leads to a host sync. This will break async scheduling

Comment (sidsingh-nvidia) [reply megatron/core/inference/sampling/flashinfer_sampling.py]: we can fallback to vllm's way of sampling in this case - https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/gpu/sample/sampler.py#L234-L242

**Comment by sidsingh-nvidia** [comment]
Can we make flashinfer the default backend (just like vllm) and gracefully fallback to torch if it is not installed?

**Review by shanmugamr1992 — APPROVED** [review]

File: megatron/core/inference/sampling/flashinfer_sampling.py (hunk 1 above)
Comment (tdene) [reply megatron/core/inference/sampling/flashinfer_sampling.py]: I recommend using https://docs.flashinfer.ai/generated/flashinfer.sampling.sampling_from_probs.html for symmetry; FlashInfer has kernels for all 4 scenarios.

Comment (santhnm2) [reply megatron/core/inference/sampling/flashinfer_sampling.py]: Done, @sidsingh-nvidia @tdene could you take another look?

**Comment by santhnm2** [comment]
> Can we make flashinfer the default backend (just like vllm) and gracefully fallback to torch if it is not installed?

Done

File: megatron/core/inference/sampling/flashinfer_sampling.py [hunk 2]
```diff
@@ -79,25 +84,62 @@ def sample_kernel(
...
-        output.copy_(
-            flashinfer.sampling.top_k_top_p_sampling_from_probs(
-                probs, top_k_safe, top_p_safe, generator=self._rng
+        if no_top_k and no_top_p:
+            # No nucleus / top-k filtering: sample the full temperature-scaled
+            # distribution. Use FlashInfer's kernel rather than torch.multinomial:
+            # multinomial forces a device-to-host sync, whereas sampling_from_probs
+            # stays on-device and keeps the RNG's philox offset advancing per launch.
+            probs = torch.softmax(scaled, dim=-1)
+            output.copy_(
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/sampling/flashinfer_sampling.py]: Do we really need these output.copy operations? Is there any reason why we cannot directly return the FI outputs?

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -1059,12 +1060,13 @@ def validate_args(args, defaults={}):
...
-                "--inference-dynamic-batching-sampling-backend=flashinfer requires "
-                "the flashinfer package; install it or pass "
+        except ImportError:
+            warnings.warn(
+                "--inference-dynamic-batching-sampling-backend=flashinfer was requested "
+                "but the flashinfer package is not installed; falling back to "
                 "--inference-dynamic-batching-sampling-backend=torch."
-            ) from e
+            )
+            args.inference_dynamic_batching_sampling_backend = 'torch'
```
Comment (sidsingh-nvidia) [inline megatron/training/arguments.py]: Suggestion to move this fallback logic to inference config so that nemoRL is also able to use it.

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

File: megatron/core/inference/sampling/flashinfer_sampling.py (hunk 2 above)
Comment (tdene) [reply megatron/core/inference/sampling/flashinfer_sampling.py]: Agreed. The `output.copy_` was there to keep static memory addresses for the graphs. If this operation is never graphed, we can always directly return.

File: megatron/training/arguments.py (hunk 3 above)
Comment (santhnm2) [reply megatron/training/arguments.py]: Removed this logic from `arguments.py`, it's already in `megatron/core/inference/config.py`

File: megatron/core/inference/sampling/flashinfer_sampling.py (hunk 2 above)
Comment (santhnm2) [reply megatron/core/inference/sampling/flashinfer_sampling.py]: Done, now directly returning

**Review by sidsingh-nvidia — APPROVED** [review]

**Review by tdene — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]

**Review by maanug-nv — APPROVED** [review]

**Comment by sidsingh-nvidia** [comment]
Pushed a commit to fix the failing test harness.

**Comment by sidsingh-nvidia** [comment]
/ok to test [b571fcd](https://github.com/NVIDIA/Megatron-LM/pull/5791/commits/b571fcd864f73986db8d7a756fb392310daaa9da)

**Review by thomasdhc — APPROVED** [review]
